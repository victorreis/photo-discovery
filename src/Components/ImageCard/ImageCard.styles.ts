import styled from 'styled-components';

import { toPx } from '../../Utils/Transform/toPx.util';
import { Typography } from '../Typography';
import { ImageCardStyleProps } from './ImageCard.types';

export const ImageCardContainer = styled.div<ImageCardStyleProps>`
  background-color: ${({ theme }) => theme.colors.background.default.normal};
  width: 10rem;
  border-radius: ${({ theme }) => toPx(theme.borders.radius.LG)};
  box-sizing: border-box;
  display: flex;
  flex-direction: column;
  box-shadow: 0 0 0.25em ${({ theme }) => theme.colors.main.effect.normal};

  &:hover {
    box-shadow: 0 0 0.75em ${({ theme }) => theme.colors.main.effect.normal};
  }
`;

export const Thumbnail = styled.img`
  width: 10rem;
  height: 10rem;
  box-sizing: border-box;
  border-top-left-radius: ${({ theme }) => toPx(theme.borders.radius.LG)};
  border-top-right-radius: ${({ theme }) => toPx(theme.borders.radius.LG)};
  cursor: pointer;
`;

export const CardTitle = styled(Typography)`
  padding: 0.5rem;
`;

export const ModalTitle = styled(Typography)`
  margin-block-end: 0;
`;
