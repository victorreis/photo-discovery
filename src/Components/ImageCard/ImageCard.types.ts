import { TestProps } from '../../Config/Tests/Test.types';

export interface RequiredImageCardProps {
  /**
   * Card identificator.
   */
  id: string;

  /**
   * Thumbnail url.
   */
  thumbnailUrl: string;

  /**
   * Image url.
   */
  imageUrl: string;

  /**
   * Title shown in the card.
   */
  title: string;
}

export interface DefaultImageCardProps {}

export interface OptionalImageCardProps {
  /**
   * Word or piece of word that will guide what words will be italicized.
   */
  italicizedWord?: string;
}

export type ImageCardProps = RequiredImageCardProps &
  DefaultImageCardProps &
  OptionalImageCardProps &
  TestProps &
  Omit<React.HTMLAttributes<HTMLDivElement>, 'children'>;

export type ImageCardStyleProps = Required<DefaultImageCardProps>;
