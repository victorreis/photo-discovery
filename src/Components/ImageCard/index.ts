export * from './ImageCard.types';
export * from './ImageCard';
