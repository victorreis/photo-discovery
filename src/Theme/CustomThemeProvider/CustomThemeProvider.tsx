import React, { useMemo, useState } from 'react';

import { ThemeProvider } from 'styled-components';

import {
  lightTheme,
  darkTheme,
  futuristicTheme,
  deepSpaceTheme,
} from '../Modes';
import { Mode, CustomThemeType } from '../Types';
import {
  ThemeContextType,
  CustomThemeProviderProps,
} from './CustomThemeProvider.types';

export const themes: Record<Mode | 'default', CustomThemeType> = {
  default: deepSpaceTheme,
  light: lightTheme,
  dark: darkTheme,
  futuristic: futuristicTheme,
  deepSpace: deepSpaceTheme,
};

export const ThemeContext = React.createContext<
  ThemeContextType | Record<string, never>
>({});

export const customThemeProviderDefaults: Required<CustomThemeProviderProps> = {
  testID: 'ThemeContextProvider',
  themeName: 'deepSpace',
};

export const CustomThemeProvider: React.FC<CustomThemeProviderProps> = (
  props
): JSX.Element => {
  const {
    testID = customThemeProviderDefaults.testID,
    children,
    themeName = customThemeProviderDefaults.themeName,
  } = props;

  const [theme, setTheme] = useState<CustomThemeType>(themes[themeName]);

  const switchTheme = (themeMode: Mode): void => {
    setTheme(themes[themeMode]);
  };

  const providerValue = useMemo(() => ({ theme, switchTheme }), [theme]);

  return (
    <ThemeContext.Provider data-testid={testID} value={providerValue}>
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    </ThemeContext.Provider>
  );
};
