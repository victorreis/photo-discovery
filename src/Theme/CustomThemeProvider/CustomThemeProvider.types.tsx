import { CustomThemeType, Mode } from '../Types';

export type ThemeContextType = {
  theme: CustomThemeType;
  switchTheme: (themeMode: Mode) => void;
};

export type CustomThemeProviderProps = { testID?: string; themeName?: Mode };
