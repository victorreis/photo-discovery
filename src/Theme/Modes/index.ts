export { darkTheme } from './Dark.theme';
export { deepSpaceTheme } from './DeepSpace.theme';
export { futuristicTheme } from './Futuristic.theme';
export { lightTheme } from './Light.theme';
