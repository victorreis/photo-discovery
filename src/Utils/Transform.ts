export const hexToRgb = (hex: string) => {
  // Expand shorthand form (e.g. "03F") to full form (e.g. "0033FF")
  const shorthandRegex = /^#?([a-f\d])([a-f\d])([a-f\d])$/i;
  const formatedHex = hex.replace(
    shorthandRegex,
    (_m, r: string, g: string, b: string) => r + r + g + g + b + b
  );

  const result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(formatedHex);
  return result
    ? `rgb(${parseInt(result[1] || '0', 16)}, ${parseInt(
        result[1] || '0',
        16
      )}, ${parseInt(result[1] || '0', 16)})`
    : 'rgb(0,0,0)';
};
