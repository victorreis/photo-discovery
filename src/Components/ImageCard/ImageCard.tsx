import { useState } from 'react';

import { TestProps } from '../../Config/Tests/Test.types';
import { Modal } from '../Modal';
import {
  CardTitle,
  ImageCardContainer,
  ModalTitle,
  Thumbnail,
} from './ImageCard.styles';
import { ImageCardProps, DefaultImageCardProps } from './ImageCard.types';

export const imageCardDefaults: Required<DefaultImageCardProps> &
  Required<TestProps> = {
  testID: 'ImageCard',
};

export const ImageCard: React.FC<ImageCardProps> = (props): JSX.Element => {
  const {
    testID = imageCardDefaults.testID,
    id,
    thumbnailUrl,
    imageUrl,
    title,
    ...others
  } = props;

  const [open, setOpen] = useState(false);

  const handleOpenClose = () => {
    setOpen((prevState) => !prevState);
  };

  return (
    <ImageCardContainer key={id} data-testid={testID} {...others}>
      <Thumbnail onClick={handleOpenClose} src={thumbnailUrl} />
      <CardTitle variant="body2">{title}</CardTitle>
      {open && (
        <Modal onClose={handleOpenClose}>
          <img alt={title} src={imageUrl} />
          <ModalTitle variant="h5">{title}</ModalTitle>
        </Modal>
      )}
    </ImageCardContainer>
  );
};
