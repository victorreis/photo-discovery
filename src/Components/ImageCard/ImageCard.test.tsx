import {
  renderJestDomCreator,
  renderRTRCreator,
  screen,
} from '../../Config/Tests/GlobalSetup.config';
import { themes } from '../../Theme/CustomThemeProvider';
import { toPx } from '../../Utils/Transform';
import { ImageCard, imageCardDefaults } from './ImageCard';
import { ImageCardProps, RequiredImageCardProps } from './ImageCard.types';

describe(`ImageCard component tests`, () => {
  const id = '1';
  const thumbnailUrl = 'https://via.placeholder.com/600/771796';
  const imageUrl = 'https://via.placeholder.com/150/771796';
  const title = 'teste';

  const requiredProps: RequiredImageCardProps = {
    id,
    thumbnailUrl,
    imageUrl,
    title,
  };

  const setup = (props?: ImageCardProps) => {
    const renderRTR = () =>
      renderRTRCreator<ImageCardProps>(ImageCard, {
        ...requiredProps,
        ...props,
      });
    const renderJestDom = () =>
      renderJestDomCreator<ImageCardProps>(ImageCard, {
        ...requiredProps,
        ...props,
      });

    return { renderRTR, renderJestDom };
  };

  describe(`behavior tests`, () => {
    it(`should render the component`, () => {
      setup().renderJestDom();
      const testInstance = screen.getByTestId(imageCardDefaults.testID);

      expect(testInstance).toBeTruthy();
    });

    it(`should render the text`, () => {
      setup().renderJestDom();
      const element = screen.getByText(title);

      expect(element).toBeInTheDocument();
    });
  });

  describe(`style tests`, () => {
    it(`should have style the Container component`, () => {
      setup().renderJestDom();
      const container = screen.getByTestId(imageCardDefaults.testID);

      expect(container).toHaveStyle({
        backgroundColor: themes.default.colors.background.default.darker,
        borderRadius: toPx(themes.default.borders.radius.MD),
      });
    });
  });

  describe(`snapshot tests`, () => {
    it(`should render correctly`, () => {
      const generatedJson = setup().renderRTR().toJSON();
      expect(generatedJson).toMatchSnapshot();
    });
  });
});
