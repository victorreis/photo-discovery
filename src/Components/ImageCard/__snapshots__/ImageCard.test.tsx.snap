// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageCard component tests snapshot tests should render correctly 1`] = `
<div
  className="ImageCardContainer-photo-discovery__sc-40sd5q-0 jIPeRU"
  data-testid="ImageCard"
>
  <img
    className="Thumbnail-photo-discovery__sc-40sd5q-1 iYbfCN"
    onClick={[Function]}
    src="https://via.placeholder.com/600/771796"
  />
  <span
    className="TypographyContainer-photo-discovery__sc-1t3hmm1-0 gqzOXP CardTitle-photo-discovery__sc-40sd5q-2 iSTYDc"
    data-testid="Typography"
  >
    teste
  </span>
</div>
`;
